tests/sterm/test_terminal.py

```diff
@@ -0,0 +1,213 @@
+
+    terminal.run("tty.122", 9600)
+
+    mock_sterm().join.assert_called_with()
+
+
+def test_closes_sterm(mock_sterm, mock_serial):
+    terminal.run("tty.122", 9600)
+
+    mock_sterm().close.assert_called_once()
+
+
+def test_closes_sterm_after_exception(mock_sterm, mock_serial):
+    mock_sterm().join.side_effect = (Exception(), None)
+    with pytest.raises(Exception):
+        terminal.run("tty.122", 9600)
+
+    mock_sterm().close.assert_called_once()
+
+
+def test_sets_terminal_special_chars(mock_serial, mock_console):
+    term = terminal.SerialTerminal(mock_serial())
+
+    assert term.exit_character == terminal.CTRL_C
+    assert term.menu_character == terminal.CTRL_T
+    assert term.reset_character == terminal.CTRL_B
+    assert term.help_character == terminal.CTRL_H
+
+
+def test_sets_terminal_rx_and_tx_encoding_to_utf8(mock_serial, mock_console):
+    term = terminal.SerialTerminal(mock_serial())
+
+    assert term.input_encoding == "UTF-8"
+    assert term.output_encoding == "UTF-8"
+
+
+def test_stops_terminal_when_ctrl_c_received(mock_serial, mock_console):
+    term = terminal.SerialTerminal(mock_serial())
+    term.alive = True
+    mock_console().getkey.return_value = terminal.CTRL_C
+
+    term.writer()
+
+    assert term.alive is False
+
+
+def test_stops_terminal_on_keyboard_interrupt(mock_serial, mock_console):
+    term = terminal.SerialTerminal(mock_serial())
+    term.alive = True
+    mock_console().getkey.side_effect = KeyboardInterrupt()
+
+    term.writer()
+
+    assert term.alive is False
+
+
+@pytest.mark.parametrize("menu_key", terminal.VALID_MENU_KEYS)
+def test_handles_valid_menu_key(menu_key, mock_serial, mock_console):
+    term = terminal.SerialTerminal(mock_serial())
+    term.handle_menu_key = mock.Mock()
+    term.alive = True
+    mock_console().getkey.side_effect = (terminal.CTRL_T, menu_key, terminal.CTRL_C)
+
+    term.writer()
+
+    term.handle_menu_key.assert_called_once_with(menu_key)
+
+
+INVALID_MENU_KEYS = tuple(set(chr(i) for i in range(0, 127)) - set(terminal.VALID_MENU_KEYS) - set([terminal.CTRL_H]))
+
+
+@pytest.mark.parametrize("menu_key", INVALID_MENU_KEYS)
+def test_ignores_invalid_menu_key(menu_key, mock_serial, mock_console):
+    term = terminal.SerialTerminal(mock_serial())
+    term.handle_menu_key = mock.Mock()
+    term.alive = True
+    mock_console().getkey.side_effect = (terminal.CTRL_T, menu_key)
+
+    with pytest.raises(StopIteration):
+        term.writer()
+
+    term.handle_menu_key.assert_not_called()
+
+
+def test_reset_sends_serial_break(mock_serial, mock_console):
+    term = terminal.SerialTerminal(mock_serial())
+
+    term.reset()
+
+    mock_serial().sendBreak.assert_called_once()
+
+
+def test_ctrl_b_sends_reset_to_serial_port(mock_serial, mock_console):
+    term = terminal.SerialTerminal(mock_serial())
+    term.alive = True
+    mock_console().getkey.side_effect = (terminal.CTRL_B, terminal.CTRL_C)
```

Comment:
Would this test pass if we did nothing for CTRL_B and made CTRL_C send a break to the serial port? How do we know it's the CTRL_B that's giving us the desired effect and not something else? I am curious why we sent CTRL_C. I guess it is to cause the console to stop blocking, but perhaps there is a more targeted way to do that.

Reply:
CTRL_C doesn't send a break to the serial port, it tells the terminal to stop running and blocking. Maybe it's better to just handle the `StopIteration` like I did in the above test, to make it clearer we're not testing CTRL_C handling. I did think the `StopIteration` handling was a bit dirty though...